hw4/04-templates.js

```diff
@@ -1,64 +1,95 @@
-app.set('view engine', 'pug');
+app.set("views", __dirname + "/views");
+app.set("view engine", "pug");
 
 // REST Countries URL
-const url = 'https://restcountries.com/v3.1/all';
+const url = "https://restcountries.com/v3.1/all";
 
-// Add your code here
+let info = [];
+let countries = [], populous = [], regions = [];
+let isPopulationinfoCalculated = false;
 
-app.get('/', (req, res) => {
+app.get("/", async (req, res) => {
   // render pug template for the index.html file
 
-  res.render('index', {
-    heading: 'Countries of the World',
-    main: 'Welcome to this application. Using the REST Countries API, we will be showing the countries and capitals of the world, the most populous countries in the world, and the number of countries in each region of the world',
+  info = await fetch(url)
+    .then(info => info.json())
+    .then(info => info)
+    .catch(e => {
+        console.log(e);
+        return [];
+      }
+    );
+
+  res.render("index", {
+    heading: "Countries of the World",
+    main: "Welcome to this application. Using the REST Countries API, we will be showing the countries and capitals of the world, the most populous countries in the world, and the number of countries in each region of the world"
   });
 });
 
-app.get('/capitals', (req, res) => {
+app.get("/capitals", (req, res) => {
   // map the output array to create an array with country names and capitals
-  // check for empty data in the output array
-
-  let countries = ['Afghanistan', 'Aland Islands', 'Albania'];
-
-  res.render('page', {
-    heading: 'Countries and Capitals',
-    results: countries,
+  // check for empty info in the output array
+  countries = info
+    .map((country, i) => `${country.name.common}  ${country.capital && country.capital[0] ? (`- ${country.capital[0]}`) : ""}`)
+    .sort()
+
+  res.render("page", {
+    heading: "Countries and Capitals",
+    results: countries
   });
 });
 
-app.get('/populous', (req, res) => {
+app.get("/populous", (req, res) => {
   // filter the output array for the countries with population of 50 million or more
   // sort the resulting array to show the results in order of population
   // map the resulting array into a new array with the country name and formatted population
 
-  let populous = ['China', 'India', 'United States of America'];
-
-  res.render('page', {
-    heading: 'Most Populous Countries',
-    results: populous,
+  if (!isPopulationinfoCalculated) {
+    info.forEach(country => {
+      if (country.population >= 50000000) {
+        populous.push(country);
+      }
+    });
+    populous = populous
+      .sort((a, b) => b.population - a.population)
+      .map(country => `${country.name.common} - ${country.population.toLocaleString()}`);
+    isPopulationinfoCalculated = true;
+  }
+
+  res.render("page", {
+    heading: "Most Populous Countries",
+    results: populous
   });
 });
 
-app.get('/regions', (req, res) => {
+app.get("/regions", (req, res) => {
```

Comment:
Love it, almost everything here is golden.
- `/capitals`  outputs with correct formatting and alphabetized
- `/populous` outputs with correct formatting (number formatting is also set correctly!) and ordered in descending population.

The only thing I noticed is that when I go to `/regions` it seems to add to the output data every time I refresh the page. I start with 6 but every time I refresh it adds the same data to the list so that it grows by 6 every time. 

Everything else here looks good though!